### PR TITLE
Introduce `SELECT MUTATION FRAGMENTS` statement

### PR DESCRIPTION
**Pull request by denesb**
SELECT MUTATION FRAGMENTS is a new select statement sub-type, which allows dumping the underling mutations making up the data of a given table. The output of this statement is mutation-fragments presented as CQL rows. Each row corresponds to a mutation-fragment. Subsequently, the output of this statement has a schema that is different than that of the underlying table.  The output schema is derived from the table's schema, as following:
* The table's partition key is copied over as-is
* The clustering key is formed from the following columns:
    - mutation_source (text): the kind of the mutation source, one of: memtable, row-cache or sstable; and the identifier of the individual mutation source.
    - partition_region (int): represents the enum with the same name.
    - the copy of the table's clustering columns
    - position_weight (int): -1, 0 or 1, has the same meaning as that in position_in_partition, used to disambiguate range tombstone changes with the same clustering key, from rows and from each other.
* The following regular columns:
    - metadata (text): the JSON representation of the mutation-fragment's metadata.
    - value (text): the JSON representation of the mutation-fragment's value.

Data is always read from the local replica, on which the query is executed. Migrating queries between coordinators is frobidden.

More details in the documentation commit (last commit).

Example:
```cql
cqlsh> CREATE TABLE ks.tbl (pk int, ck int, v int, PRIMARY KEY (pk, ck));

cqlsh> DELETE FROM ks.tbl WHERE pk = 0;                                                                                                                                                                                                                       
cqlsh> DELETE FROM ks.tbl WHERE pk = 0 AND ck > 0 AND ck < 2;                                                                                                                                                                                                 
cqlsh> INSERT INTO ks.tbl (pk, ck, v) VALUES (0, 0, 0);                                                                                                                                                                                                       
cqlsh> INSERT INTO ks.tbl (pk, ck, v) VALUES (0, 1, 0);                                                                                                                                                                                                       
cqlsh> INSERT INTO ks.tbl (pk, ck, v) VALUES (0, 2, 0);                                                                                                                                                                                                       
cqlsh> INSERT INTO ks.tbl (pk, ck, v) VALUES (1, 0, 0);                                                                                                                                                                                                       
cqlsh> SELECT * FROM ks.tbl;

 pk | ck | v                                                                                                                                                                                                                                                  
----+----+---                                                                                                                                                                                                                                                 
  1 |  0 | 0                                                                                                                   
  0 |  0 | 0                                                                                                                                                                                                                                                  
  0 |  1 | 0                                                                                                                   
  0 |  2 | 0                                                                                                                                                                                                                                                  
                                                                                                                               
(4 rows) 
cqlsh> SELECT * FROM MUTATION_FRAGMENTS(ks.tbl);                                                                                                                                                                                                              
                                                                                                                                                                                                                                                              
 pk | mutation_source | partition_region | ck | position_weight | metadata                                                                                                                 | mutation_fragment_kind | value                                   
----+-----------------+------------------+----+-----------------+--------------------------------------------------------------------------------------------------------------------------+------------------------+-----------                              
  1 |      memtable:0 |                0 |    |                 |                                                                                                         {"tombstone":{}} |        partition start |      null
  1 |      memtable:0 |                2 |  0 |               0 | {"marker":{"timestamp":1688122873341627},"columns":{"v":{"is_live":true,"type":"regular","timestamp":1688122873341627}}} |         clustering row | {"v":"0"}                               
  1 |      memtable:0 |                3 |    |                 |                                                                                                                     null |          partition end |      null
  0 |      memtable:0 |                0 |    |                 |                                      {"tombstone":{"timestamp":1688122848686316,"deletion_time":"2023-06-30 11:00:48z"}} |        partition start |      null                               
  0 |      memtable:0 |                2 |  0 |               0 | {"marker":{"timestamp":1688122860037077},"columns":{"v":{"is_live":true,"type":"regular","timestamp":1688122860037077}}} |         clustering row | {"v":"0"}
  0 |      memtable:0 |                2 |  0 |               1 |                                      {"tombstone":{"timestamp":1688122853571709,"deletion_time":"2023-06-30 11:00:53z"}} | range tombstone change |      null                               
  0 |      memtable:0 |                2 |  1 |               0 | {"marker":{"timestamp":1688122864641920},"columns":{"v":{"is_live":true,"type":"regular","timestamp":1688122864641920}}} |         clustering row | {"v":"0"}
  0 |      memtable:0 |                2 |  2 |              -1 |                                                                                                         {"tombstone":{}} | range tombstone change |      null                               
  0 |      memtable:0 |                2 |  2 |               0 | {"marker":{"timestamp":1688122868706989},"columns":{"v":{"is_live":true,"type":"regular","timestamp":1688122868706989}}} |         clustering row | {"v":"0"}                               
  0 |      memtable:0 |                3 |    |                 |                                                                                                                     null |          partition end |      null                               

(10 rows)
```

Perf simple query:
```
/build/release/scylla perf-simple-query -c1 -m2G --duration=60
```

Before:
```
median 141596.39 tps ( 62.1 allocs/op,  13.1 tasks/op,   43688 insns/op,        0 errors)
median absolute deviation: 137.15
maximum: 142173.32
minimum: 140492.37
```
After:
```
median 141889.95 tps ( 62.1 allocs/op,  13.1 tasks/op,   43692 insns/op,        0 errors)
median absolute deviation: 167.04
maximum: 142380.26
minimum: 141025.51
```

Fixes: https://github.com/scylladb/scylladb/issues/11130

### CODE REVIEW

**Comment by denesb** [comment]
This PR is a minimal POC of the presented approach, opened as an RFC to get early feedback on the approach.
I tested only the simplest of cases manually (see example above).
This is a continuation of the now closed https://github.com/scylladb/scylladb/pull/14083, the approach used in which proved to be a dead-end.

File: test/cql-pytest/test_select_mutation_fragments.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+
+@pytest.fixture(scope="module")
+def test_table(cql, test_keyspace):
+    """ Prepares a table for the mutation dump tests to work with."""
+    with util.new_test_table(cql, test_keyspace, 'pk1 int, pk2 int, ck1 int, ck2 int, v text, s text static, PRIMARY KEY ((pk1, pk2), ck1, ck2)') as table:
+        yield table
+
+
+def test_mutation_dump_order_by(cql, test_table):
```
Comment (nyh) [inline test/cql-pytest/test_select_mutation_fragments.py:22]: I know this is just an early version, but eventually please add tests also for the success cases. For example, you can use `nodetool.flush()` and `BYPASS CACHE` to ensure that your data comes from sstable, not memtable (I don't know how to make sure that you get something from a memtable or cache, you can be pretty sure, but is there a way to be 100% sure).

Also please add some documentation for this feature, perhaps in docs/cql/cql-extensions.md if you think real users might want to use it, or docs/dev/cql-extensions-internal.md if you think real users shouldn't use it and it's mostly for development and debugging.

Comment (denesb) [reply test/cql-pytest/test_select_mutation_fragments.py:22]: Yes, I do plan to add much more tests, both positive and negative. I accidentally pushed my "dev" version to this PR (I do push quite often, as a way to save my progress and I usually do that to a separate branch, to not pollute the PR).
I will also add extensive documentation for this. Thanks for reminding me of `cql-extenstions.md`, I forgot about that.

**Comment by avikivity** [comment]
We need to disambiguate the syntax. It's too close to SELECT mutation, fragments FROM foo.

This is an developer oriented query.

It's also very different from regular SELECT - it goes directly to the replica (is this correct?), not through coordination.

**Comment by avikivity** [comment]
Let's change partition_region to be text, it's more readable.

File: replica/table.cc [hunk 2]
```diff
@@ -350,6 +350,14 @@ flat_mutation_reader_v2 table::make_memtable_reader(schema_ptr schema, reader_pe
     return make_combined_reader(std::move(schema), std::move(permit), std::move(readers), streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
 }
 
+flat_mutation_reader_v2 table::make_cache_reader(schema_ptr schema, reader_permit permit, const dht::partition_range& range,
+        const query::partition_slice& slice, tracing::trace_state_ptr ts) {
```
Comment (avikivity) [inline replica/table.cc:355]: Please keep nonpopulating in the name. It's a critical aspect of the function.

File: replica/mutation_dump.cc [hunk 3]
```diff
@@ -0,0 +1,366 @@
+/*
+ * Modified by ScyllaDB
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
```
Comment (avikivity) [inline replica/mutation_dump.cc:7]: This isn't Apache-derived code.

File: service/storage_proxy.hh [hunk 4]
```diff
@@ -101,6 +101,37 @@ using allow_hints = bool_class<allow_hints_tag>;
...
+        , trace_state(std::move(trace_state))
+        , preferred_replicas(std::move(preferred_replicas))
+        , read_repair_decision(read_repair_decision) {
+    }
+
+    storage_proxy_clock_type::time_point timeout(storage_proxy& sp) const {
+        return _timeout;
+    }
+};
+
```
Comment (avikivity) [inline service/storage_proxy.hh:134]: Not this patch's fault, but these options-style things are better as structs, not classes. But the series is right not to touch it.

File: service/pager/query_pager.hh [hunk 5]
```diff
@@ -21,10 +21,19 @@
...
 namespace pager {
 
+using query_function = std::function<future<exceptions::coordinator_result<service::storage_proxy_coordinator_query_result>>(
+        service::storage_proxy& sp,
+        schema_ptr schema,
+        lw_shared_ptr<query::read_command> cmd,
+        dht::partition_range_vector&& partition_ranges,
+        db::consistency_level cl,
+        service::storage_proxy_coordinator_query_options optional_params)>;
+
```
Comment (avikivity) [inline service/pager/query_pager.hh:36]: Another way to do it is to pass a different storage_proxy that talks to the local replica. We could make this part of the Great Storage Proxy Refactoring. /cc @xemul @kbr-scylla. Using this function is a good solution for now.

File: service/pager/query_pagers.cc [hunk 6]
```diff
@@ -57,7 +58,21 @@ query_pager::query_pager(service::storage_proxy& p, schema_ptr s,
...
+                service::storage_proxy& sp,
+                schema_ptr schema,
+                lw_shared_ptr<query::read_command> cmd,
+                dht::partition_range_vector&& partition_ranges,
+                db::consistency_level cl,
+                service::storage_proxy_coordinator_query_options optional_params) {
+            return sp.query_result(std::move(schema), std::move(cmd), std::move(partition_ranges), cl, std::move(optional_params));
+        };
+    }
+}
```
Comment (avikivity) [inline service/pager/query_pagers.cc:75]: Please run perf_simple_query to see this doesn't generate some silly regression.

File: cql3/Cql.g [hunk 7]
```diff
@@ -389,6 +389,7 @@ selectStatement returns [std::unique_ptr<raw::select_statement> expr]
     }
     : K_SELECT (
                 ( K_JSON { statement_subtype = raw::select_statement::parameters::statement_subtype::JSON; } )?
+                ( K_MUTATION K_FRAGMENTS { statement_subtype = raw::select_statement::parameters::statement_subtype::MUTATION_FRAGMENTS; } )?
                 ( K_DISTINCT { is_distinct = true; } )?
```
Comment (avikivity) [inline cql3/Cql.g:393]: Another idea: mutilate the FROM clause.

```cql
SELECT *
FROM mutation_fragments(tab)
WHERE ...
```

**Comment by denesb** [comment]
> Let's change partition_region to be text, it's more readable.

If I make it text, it won't sort the right way anymore (it is part of the clustering key). I also can't remove it as it makes sure the partition-end is sorted to the very end. I can remove the partition-end altogether (it is there just for completeness' sake), but we still have partition-start and static-row with the same clustering key, without partition-region.

File: cql3/restrictions/statement_restrictions.cc [hunk 8]
```diff
@@ -528,9 +534,13 @@ bool statement_restrictions::has_eq_restriction_on_column(const column_definitio
...
     if (need_filtering()) {
+        std::optional<secondary_index::index> opt_idx;
+      try {
         auto cf = db.find_column_family(_schema);
         auto& sim = cf.get_index_manager();
-        auto opt_idx = std::get<0>(find_idx(sim));
+        opt_idx = std::get<0>(find_idx(sim));
+      } catch (data_dictionary::no_such_column_family&) {
+      }
         auto column_uses_indexing = [&opt_idx] (const column_definition* cdef, const expr::expression* single_col_restr) {
```
Comment (avikivity) [inline cql3/restrictions/statement_restrictions.cc:544]: Again this will put a big question mark on a future reader of the code's face.

**Comment by denesb** [comment]
> We need to disambiguate the syntax. It's too close to SELECT mutation, fragments FROM foo.
> 
> This is an developer oriented query.
> 
> It's also very different from regular SELECT - it goes directly to the replica (is this correct?), not through coordination.

How about `SELECT MUTATION_FRAGMENTS`?
Note that we have precedent on adding things after `SELECT`: `SELECT DISTINCT` and `SELECT JSON`.

File: test/cql-pytest/test_select_mutation_fragments.py (hunk 1 above)
Comment (avikivity) [reply test/cql-pytest/test_select_mutation_fragments.py:22]: We should focus more on the in-tree reference documentation (@annastuchlik). cql-extensions.md is a good place to list the extensions, but the real documentation should be in the reference.

Make sure to slap an "unstable interface, may change at any moment". It's not useful for applications, but people will manage to depend on it.

File: cql3/Cql.g (hunk 7 above)
Comment (denesb) [reply cql3/Cql.g:393]: This could also work. But note that I also want a working `WHERE` clause. One of the big reasons why I stuck with CQL for this feature instead of API, is the built-in data manipulation capabilities in CQL. I expect that in the vast majority of cases, we will not want to read everything from this, but restrict it in different ways. E.g. read partition x from memtable, read all partitions from cache, read only tombstones from cache, etc.

File: cql3/restrictions/statement_restrictions.cc [hunk 9]
```diff
@@ -380,18 +380,24 @@ statement_restrictions::statement_restrictions(data_dictionary::database db,
...
             && !type.is_delete();
     _has_queriable_pk_index = parition_key_restrictions_have_supporting_index(sim, allow_local)
             && !type.is_delete();
     _has_queriable_regular_index = expr::index_supports_some_column(_nonprimary_key_restrictions, sim, allow_local)
             && !type.is_delete();
+  } catch (data_dictionary::no_such_column_family&) {
+    _has_queriable_ck_index = false;
+    _has_queriable_pk_index = false;
+    _has_queriable_regular_index = false;
+  }
```
[the remark this answers is not in the document]
Comment (denesb) [reply cql3/restrictions/statement_restrictions.cc:400]: Yes, this was a quick hack to get this query working. I will pass down a parameter to this method instead. Indeed, as-is, this can cover up real problems.

**Comment by avikivity** [comment]
> > We need to disambiguate the syntax. It's too close to SELECT mutation, fragments FROM foo.
> > This is an developer oriented query.
> > It's also very different from regular SELECT - it goes directly to the replica (is this correct?), not through coordination.
> 
> How about `SELECT MUTATION_FRAGMENTS`? 


```cql
CREATE TABLE foo (MUTATION_FRAGMENTS int PRIMARY KEY);

SELECT MUTATION_FRAGMENTS FROM foo;
```


> Note that we have precedent on adding things after `SELECT`: `SELECT DISTINCT` and `SELECT JSON`.

This is not a mainstream extension like SELECT JSON or part of the SQL standard like SELECT DISTINCT. We have to be more careful.

**Comment by denesb** [comment]
> It's also very different from regular SELECT - it goes directly to the replica (is this correct?), not through coordination.

Correct. Although I will have to implement an RPC, because the driver will not know to send all queries to the same coordinator. So we will have quasi coordination, but not through storage-proxy, because storage proxy only works with known tables. This query is equivalent of creating a temporary virtual table, which derives from the underlying table. Maybe that would be the correct way to expose this to CQL? But this really does sound like a project. I suspect in the circumstances where this query would be used, the last thing we want is a DDL on the cluster (even if temporary).

**Comment by avikivity** [comment]
> > It's also very different from regular SELECT - it goes directly to the replica (is this correct?), not through coordination.
> 
> Correct. Although I will have to implement an RPC, because the driver will not know to send all queries to the same coordinator. So we will have quasi coordination, but not through storage-proxy, because storage proxy only works with known tables. This query is equivalent of creating a temporary virtual table, which derives from the underlying table. Maybe that would be the correct way to expose this to CQL? But this really does sound like a project. I suspect in the circumstances where this query would be used, the last thing we want is a DDL on the cluster (even if temporary).

Isn't there a way to restrict the driver to one node? Perhaps we can hack it in. cqlsh for administration will also require it.

**Comment by denesb** [comment]
> Isn't there a way to restrict the driver to one node? Perhaps we can hack it in. cqlsh for administration will also require it.

There is, we use it in dtest. With the target audience in mind (engineers) do you think we can trust them to know how to set this up (will they read the manual)? I would certainly prefer no to complicate this with an RPC call (I haven't implemented it yet).

**Comment by denesb** [comment]
To avoid nasty suprises, I can put the address of the node who started to serve this query into the paging-state's "last_replicas" field. Then I can check this field and abort the query if this field doesn't contain the current node's address.

**Comment by avikivity** [comment]
> > Isn't there a way to restrict the driver to one node? Perhaps we can hack it in. cqlsh for administration will also require it.
> 
> There is, we use it in dtest. With the target audience in mind (engineers) do you think we can trust them to know how to set this up (will they read the manual)? I would certainly prefer no to complicate this with an RPC call (I haven't implemented it yet).

Maybe we can make it simple/easy. Like setting up a local listener and having the admin cqlsh use it.

**Comment by denesb** [comment]
> Maybe we can make it simple/easy. Like setting up a local listener and having the admin cqlsh use it.

Ok, I will push out the RPC implementation as the last item. We can think about it some more in the meanwhile.

File: cql3/Cql.g (hunk 7 above)
Comment (denesb) [reply cql3/Cql.g:393]: Ah, I missed that you talked about the `FROM` clause, not the `WHERE` clause.
Let me try to see how a "mutilated" `FROM` this would look like. I like that it emphasizes that we are doing something to the entire table.

Comment (denesb) [reply cql3/Cql.g:393]: This was quite easy:
```cql
cqlsh> select * from mutation_fragments(ks.tbl) where pk = 0;

 pk | mutation_source_kind | partition_region | ck | position_weight | mutation_fragment_kind | value
----+----------------------+------------------+----+-----------------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  0 |              sstable |                0 |    |            null |        partition start |                                                                                                                                                                                                                                           null
  0 |              sstable |                2 |  0 |               0 |         clustering row | {mutation_fragment: clustering row {position: clustered, ckp{000400000000}, 0} {clustering_row: ck 000400000000 dr {deletable_row: {row_marker: 1687357831647535 0 0} {{row:\n    v{ atomic_cell{0,ts=1687357831647535,expiry=-1,ttl=0} }}}}}}
  0 |              sstable |                2 |  1 |               0 |         clustering row | {mutation_fragment: clustering row {position: clustered, ckp{000400000001}, 0} {clustering_row: ck 000400000001 dr {deletable_row: {row_marker: 1687357833978414 0 0} {{row:\n    v{ atomic_cell{0,ts=1687357833978414,expiry=-1,ttl=0} }}}}}}
  0 |              sstable |                2 |  2 |               0 |         clustering row | {mutation_fragment: clustering row {position: clustered, ckp{000400000002}, 0} {clustering_row: ck 000400000002 dr {deletable_row: {row_marker: 1687357835597943 0 0} {{row:\n    v{ atomic_cell{0,ts=1687357835597943,expiry=-1,ttl=0} }}}}}}
  0 |              sstable |                3 |    |            null |          partition end |                                                                                                                                                                                                                                           null

(5 rows)
```
What do you think?

Comment (denesb) [reply cql3/Cql.g:393]: Here is the diff I had to do (compared to the current version):
```diff
diff --git a/cql3/Cql.g b/cql3/Cql.g
index fcd8d75ce4..1a2eb2205e 100644
--- a/cql3/Cql.g
+++ b/cql3/Cql.g
@@ -389,11 +389,13 @@ selectStatement returns [std::unique_ptr<raw::select_statement> expr]
     }
     : K_SELECT (
                 ( K_JSON { statement_subtype = raw::select_statement::parameters::statement_subtype::JSON; } )?
-                ( K_MUTATION K_FRAGMENTS { statement_subtype = raw::select_statement::parameters::statement_subtype::MUTATION_FRAGMENTS; } )?
                 ( K_DISTINCT { is_distinct = true; } )?
                 sclause=selectClause
                )
-      K_FROM cf=columnFamilyName
+      K_FROM (
+                cf=columnFamilyName
+                | K_MUTATION_FRAGMENTS '(' cf=columnFamilyName ')' { statement_subtype = raw::select_statement::parameters::statement_subtype::MUTATION_FRAGMENTS; }
+             )
       ( K_WHERE w=whereClause { wclause = std::move(w); } )?
       ( K_GROUP K_BY gbcolumns=listOfIdentifiers)?
       ( K_ORDER K_BY orderByClause[orderings] ( ',' orderByClause[orderings] )* )?
@@ -2069,8 +2071,7 @@ basic_unreserved_keyword returns [sstring str]
         | K_DESCRIBE
         | K_DESC
         | K_EXECUTE
-        | K_MUTATION
-        | K_FRAGMENTS
+        | K_MUTATION_FRAGMENTS
         ) { $str = $k.text; }
     ;
 
@@ -2279,8 +2280,7 @@ K_PRUNE:       P R U N E;
 
 K_EXECUTE:     E X E C U T E;
 
-K_MUTATION:    M U T A T I O N;
-K_FRAGMENTS:   F R A G M E N T S;
+K_MUTATION_FRAGMENTS:    M U T A T I O N '_' F R A G M E N T S;
 
 // Case-insensitive alpha characters
 fragment A: ('a'|'A');
```

Comment (denesb) [reply cql3/Cql.g:393]: I think I like this version better, and there is less chance of mixing it up with columns, as you pointed out.

**Comment by denesb** [comment]
v2:
* New syntax: `SELECT * FROM MUTATION_FRAGMENTS($table)` instead of `SELECT MUTATION FRAGMENTS FROM $table`;
* Port over tests from previous PR;
* Add some more tests (much more to come);
* Range scan support;
* Sharding support;

Not all comments and not all TODOs are addressed yet. I'm publishing this version mainly to get the updated syntax out there for review.

File: test/cql-pytest/test_select_mutation_fragments.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+# Copyright 2023-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+# Tests concerning the SELECT MUTATION FRAGMENT statement, which allows dumping
```
Comment (nyh) [inline test/cql-pytest/test_select_mutation_fragments.py:6]: Don't forget to change the references here (and in commit messages and perhaps other places) to the new syntax.

File: test/cql-pytest/test_select_mutation_fragments.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    def expect_sources(*expected_sources):
+        for src in ('memtable', 'row-cache', 'sstable'):
+            rows = list(cql.execute(f"SELECT * FROM MUTATION_FRAGMENTS({test_table}) WHERE pk1 = {pk1} AND pk2 = {pk2} AND mutation_source_kind = '{src}'"))
+            if src in expected_sources:
+                assert len(rows) == 3 # partition-start, clustering-row, partition-end
+            else:
+                assert len(rows) == 0
+
+    cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
+    expect_sources('memtable')
```
Comment (nyh) [inline test/cql-pytest/test_select_mutation_fragments.py:90]: Can we be sure that because of some background event or the phase of the moon or something, Scylla doesn't flush the memtable without our intervention, I realize that it's very unlikely, but I would feel better if you could force Scylla *not* to flush this memtable. Do we have some sort of nodetool command or something to prevent flushing? (I know in the past we had something like this with "in-memory tables")

File: test/cql-pytest/test_select_mutation_fragments.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+
+@pytest.fixture(scope="module")
+def test_table(cql, test_keyspace):
+    """ Prepares a table for the mutation dump tests to work with."""
+    with util.new_test_table(cql, test_keyspace, 'pk1 int, pk2 int, ck1 int, ck2 int, v text, s text static, PRIMARY KEY ((pk1, pk2), ck1, ck2)') as table:
+        yield table
+
+
+def test_smoke(cql, test_table):
```
Comment (nyh) [inline test/cql-pytest/test_select_mutation_fragments.py:24]: It seems that all tests in this file are Scylla-only because they test a feature that doesn't exist in Cassandra. Please either add a "scylla_only" to each test, or perhaps even better in this case, add a auto-use fixture which will add scylla_only to each and every test in this file. Something like (unchecked):
```
@pytest.fixture(scope="function", autouse=True)
def all_scylla_only(scylla_only):
    pass
```

The idea is that `test/cql-pytest/run-cassandra test_select_mutation_fragments.py` will skip all these tests.

File: cql3/Cql.g (hunk 7 above)
Comment (avikivity) [reply cql3/Cql.g:392]: Yes

File: test/cql-pytest/test_select_mutation_fragments.py (hunk 11 above)
Comment (denesb) [reply test/cql-pytest/test_select_mutation_fragments.py:90]: I'm not aware of any mechanism that force-prevents memtable flushing. But I also don't know of any way scylla would flush memtables, other than reaching the memory limit.

File: cql3/Cql.g (hunk 7 above)
Comment (bhalevy) [reply cql3/Cql.g:392]: FWIW, `SELECT mutation_fragments(*) FROM tab` makes even more sense to me, since returning mutation_fragments instead of the actual data is a function/transformation of the returned data.
With that you can also restrict the columns you want mutation fragments for, e.g. `SELECT mutation_fragments(my, columns) FROM tab WHERE my_row = 'this_row' LIMIT 10`.

File: test/cql-pytest/test_select_mutation_fragments.py (hunk 11 above)
Comment (avikivity) [reply test/cql-pytest/test_select_mutation_fragments.py:90]: Preventing flushes = preventing new writes

Memtables can be flushed for different reasons, like running out of memtable memory, running out of commitlog space, explicit request, or moving a tablet.

File: cql3/Cql.g (hunk 7 above)
Comment (denesb) [reply cql3/Cql.g:392]: Currently the output schema of this statement is not that of the underlying table. So you cannot put columns of the underlying table in the selector, you can only use columns of the output schema, so what you propose above doesn't work.

Comment (avikivity) [reply cql3/Cql.g:392]: Yes, a straightforward reading of SELECT mutation_fragments(*) means "transform all columns of the table into one column", but that's not what we want.

The new syntax means "transform the table into some view, then select all columns from the view.

In that sense, SELECT MUTATION FRAGMENTS * FROM tab would also work, since "MUTATION FRAGMENTS" modifies the SELECT, not the *.

**Comment by avikivity** [comment]
> > > We need to disambiguate the syntax. It's too close to SELECT mutation, fragments FROM foo.
> > > This is an developer oriented query.
> > > It's also very different from regular SELECT - it goes directly to the replica (is this correct?), not through coordination.
> > 
> > 
> > How about `SELECT MUTATION_FRAGMENTS`?
> 
> ```sql
> CREATE TABLE foo (MUTATION_FRAGMENTS int PRIMARY KEY);
> 
> SELECT MUTATION_FRAGMENTS FROM foo;
> ```
> 
> > Note that we have precedent on adding things after `SELECT`: `SELECT DISTINCT` and `SELECT JSON`.
> 
> This is not a mainstream extension like SELECT JSON or part of the SQL standard like SELECT DISTINCT. We have to be more careful.

`SELECT MUTATION FRAGMENTS *|selectors` is viable, since it's gramatically different.

**Comment by denesb** [comment]
v3:
* Split mutation-source-kind into individual mutation sources
* Enforce sticky coordinators (query cannot be migrated between coordinators)
* Replace hacks in `cql3/restrictions/statement_restrictions.cc` with a proper solution
* Many-many bugs fixed (found by the new tests)
* Tests
* Docs

This is now feature-complete now, from my POV.

There is one question to resolve yet: range tombstones. My goal with this statement was to present the underlying mutation-fragment stream as-is. Range tombstones however cause problems, because they can be generated on clustering range boundaries. In particular, in the case of a paged query, a range tombstone change will be generated at the start and end of each page, if there is an active range tombstone change. Not sure what do do about this. Should I make an effort to filter these out, or should I just present the stream as-is. I have an xfailing range tombstone test, because of this.

**Comment by denesb** [comment]
@annastuchlik please review the documentation part (lat commit).

**Comment by annastuchlik** [comment]
@denesb The documentation for this feature is solid and exhaustive. I only have some suggestions for language improvements, but before I add them, I'd like to discuss the location of the document you created.
I think it doesn't belong to "Admin Tools," as the tools that go into that category are of a different nature.

This feature extends the SELECT statement, so I think it should go under CQL.
How about adding it to https://opensource.docs.scylladb.com/stable/cql/cql-extensions.html? Does it make sense to you?

Having said that, I'm wondering if it should even be added to customer-facing documentation. It is not final and may break something; also, the doc says that the audience should be "familiar with the internals of ScyllaDB". That doesn't sound like user documentation. 

Perhaps it should be documented internally in https://github.com/scylladb/scylladb/tree/master/docs/dev and linked from https://opensource.docs.scylladb.com/stable/cql/cql-extensions.html.
What do you think?

File: replica/mutation_dump.hh [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+#include "query-result.hh"
+
+namespace replica::mutation_dump {
+
+schema_ptr generate_output_schema_from_underlying_schema(schema_ptr underlying_schema);
+
+future<foreign_ptr<lw_shared_ptr<query::result>>> dump_mutations(
+        sharded<database>& db,
+        schema_ptr output_schema,
+        schema_ptr underlying_schema, // must have been generated from `output_schema`, with `generate_output_schema_from_underlying_schema()`
```
Comment (avikivity) [inline replica/mutation_dump.hh:21]: Comment is confused, it's the output schema that must have been generated from the underlying schema.

File: replica/mutation_dump.cc [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+    sb.with_column("partition_region", byte_type, column_kind::clustering_key);
+    for (const auto& ck_col : underlying_schema->clustering_key_columns()) {
+        sb.with_column(ck_col.name(), ck_col.type, column_kind::clustering_key);
+    }
+    sb.with_column("position_weight", byte_type, column_kind::clustering_key);
+    // regular columns
+    sb.with_column("mutation_fragment_kind", utf8_type);
+    sb.with_column("metadata", utf8_type);
+    sb.with_column("value", utf8_type);
+    return sb.build();
```
Comment (avikivity) [inline replica/mutation_dump.cc:476]: I think we should set the uuid as a transformation of the input table_schema_version (say xor the lower bits with some pattern). This way, g_o_s_f_u_s() always returns the same schema if given the same input. May need to stop randomization of the returned schema's table_schema_version. Should test it too - calling gosfus twice on the same input should result in identical, not just equivalent schemas.

**Comment by avikivity** [comment]
Very good.

File: replica/mutation_dump.cc (hunk 14 above)
Comment (avikivity) [reply replica/mutation_dump.cc:476]: Or maybe better: transform table_id into table_id, and version into version.

**Comment by annastuchlik** [comment]
> We already have expert tools under the `Admin Tools` section, like sstabledump, scylla sstable, diagnostics and so on. These are all aimed at experts who are familiar with scylla internals. That said, this new tool might be the most expert oriented tool yet.

@denesb OK, I understand it should be part of user-facing documentation rather than dev docs.
I still think, though, it doesn't belong to the Admin Tools section, but we can leave it as is now. There's an [issue ](https://github.com/scylladb/scylladb/issues/14583)to reorganize that section, so I may need to move it somewhere else anyway.

**Review by annastuchlik — APPROVED** [review]

File: replica/mutation_dump.cc (hunk 14 above)
Comment (denesb) [reply replica/mutation_dump.cc:558]: Done in v4.

File: replica/mutation_dump.hh (hunk 13 above)
Comment (denesb) [reply replica/mutation_dump.hh]: Fixed in v4.

**Comment by nyh** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2549/

FAILED test/topology_experimental_raft/test_tablets.py::test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode

I'm trying to gather more information to create an issue. Understanding test.py failures was made difficult by having a large number of different log files without links and I'm not test.py-savvy enough to understand where to look.

**Comment by nyh** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2549/

I couldn't find a similar failure so I opened https://github.com/scylladb/scylladb/issues/14746

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2554/

One of my tests is flaky.

**Comment by denesb** [comment]
v4.1:
* Fix flaky test `test_select_from_mutation_fragments.py::test_ck_in_query` by disabling auto-compaction. The test wants to select data from a specific sstable and needs this sstable to not dissapear due to compaction, while the test is running.

**Comment by avikivity** [comment]
> v4:
> 
> * Generate the transformed schema's id and version deterministically, from that of the underlying schema.


If we ever need to cross-node query this view, this determinism will help.
